docs/source/cpp/streaming_execution.rst

```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+   * - ``scan``
+     - :class:`arrow::compute::ScanNodeOptions` 
+   * - ``hash_join``
+     - :class:`arrow::compute::HashJoinNodeOptions`
+   * - ``write``
+     - :class:`arrow::dataset::WriteNodeOptions`
+   * - ``union``
+     - N/A
+
+
+.. _stream_execution_source_docs:
+
+``source``
+----------
+
+A `source` operation can be considered as an entry point to create a streaming execution plan. 
+:class:`arrow::compute::SourceNodeOptions` are used to create the ``source`` operation.  The
+`source` operation is the most generic and flexible type of source currently available but it can
+be quite tricky to configure.  To process data from files the scan operation is likely a simpler choice.
+The source node requires some kind of function that can be called to poll for more data.  This
+function should take no arguments and should return an
+``arrow::Future<std::shared_ptr<arrow::util::optional<arrow::RecordBatch>>>``.
+This function might be reading a file, iterating through an in memory structure, or receiving data
+from a network connection.  The arrow library refers to these functions as `arrow::AsyncGenerator`
+and there are a number of utilities for working with these functions.  For this example we use 
+a vector of record batches that we've already stored in memory.
+In addition, the schema of the data must be known up front.  Arrow's streaming execution
+engine must know the schema of the data at each stage of the execution graph before any
+processing has begun.  This means we must supply the schema for a source node separately
+from the data itself.
+
+Struct to hold the data generator definition:
+
+.. literalinclude:: ../../../cpp/examples/arrow/execution_plan_documentation_examples.cc
+  :language: cpp
+  :start-after: (Doc section: BatchesWithSchema Definition)
+  :end-before: (Doc section: BatchesWithSchema Definition)
+  :linenos:
+  :lineno-match:
+
+Generating sample Batches for computation:
+
+.. literalinclude:: ../../../cpp/examples/arrow/execution_plan_documentation_examples.cc
+  :language: cpp
+  :start-after: (Doc section: MakeBasicBatches Definition)
+  :end-before: (Doc section: MakeBasicBatches Definition)
+  :linenos:
+  :lineno-match:
+
+Example of using ``source`` (usage of sink is explained in detail in :ref:`sink<stream_execution_sink_docs>`):
+
+.. literalinclude:: ../../../cpp/examples/arrow/execution_plan_documentation_examples.cc
+  :language: cpp
+  :start-after: (Doc section: Source Example)
+  :end-before: (Doc section: Source Example)
+  :linenos:
+  :lineno-match:
+
+.. _stream_execution_filter_docs:
+
+``filter``
+----------
+
+``filter`` operation, as the name suggests, provides an option to define data filtering
+criteria. It selects rows matching a given expression. 
+Filters can be written using :class:`arrow::compute::Expression`. 
+For example, if we wish to keep rows where the value of column ``b``
+is greater than 3,  then we can use the following expression:: 
```

Comment:
Rather a small nit, but can we consistently line-wrap these paragraphs? Text editors should have a shortcut to do this automatically.